# EBAC - Regressão II - regressão múltipla

## Tarefa I

#### Previsão de renda

Vamos trabalhar com a base 'previsao_de_renda.csv', que é a base do seu próximo projeto. Vamos usar os recursos que vimos até aqui nesta base.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

In [213]:
import pandas as pd
import patsy
from patsy import dmatrix
import statsmodels.api as sm

In [214]:
df = pd.read_csv('previsao_de_renda.csv')

In [215]:
df["qt_pessoas_residencia"] = df["qt_pessoas_residencia"].astype(int)
df["tempo_emprego"] = df["tempo_emprego"].fillna(0)
df["tempo_emprego"] = df["tempo_emprego"].astype(int)
df.head()

Unnamed: 0    data_ref  id_cliente sexo  posse_de_veiculo  posse_de_imovel  \
0           0  2015-01-01       15056    F             False             True   
1           1  2015-01-01        9968    M              True             True   
2           2  2015-01-01        4312    F              True             True   
3           3  2015-01-01       10639    F             False             True   
4           4  2015-01-01        7064    M              True            False   

   qtd_filhos        tipo_renda           educacao estado_civil  \
0           0        Empresário         Secundário     Solteiro   
1           0       Assalariado  Superior completo       Casado   
2           0        Empresário  Superior completo       Casado   
3           1  Servidor público  Superior completo       Casado   
4           0       Assalariado         Secundário     Solteiro   

  tipo_residencia  idade  tempo_emprego  qt_pessoas_residencia    renda  
0            Casa     26              6                      1  8060.34  
1            Casa     28              7                      2  1852.15  
2            Casa     35              0                      2  2253.89  
3            Casa     30              4                      3  6600.77  
4   Governamental     33              4                      1  6475.97

1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
    - Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
    - Mantenha sempre a categoria mais frequente como casela de referência
    - Avalie os parâmetros e veja se parecem fazer sentido prático.  


2. Remova a variável menos significante e analise:
    - Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
    - Observe os parâmetros e veja se algum se alterou muito.  


3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor. 
    

In [216]:
max_qtd_filhos = df["qtd_filhos"].value_counts().idxmax()
max_idade = df["idade"].value_counts().idxmax()
max_tempo_emprego = df["tempo_emprego"].value_counts().idxmax()
max_qt_pessoas_residencia = df["qt_pessoas_residencia"].value_counts().idxmax()
max_qtd_filhos, max_idade, max_tempo_emprego, max_qt_pessoas_residencia

(0, 40, 0, 2)

In [217]:
# variaveis qualitativas como dummies
y, x = patsy.dmatrices(f"""renda ~ C(idade, Treatment({max_idade}))
                        + C(qtd_filhos, Treatment({max_qtd_filhos}))
                        + C(tempo_emprego, Treatment({max_tempo_emprego}))
                        + C(qt_pessoas_residencia, Treatment({max_qt_pessoas_residencia}))
                        + sexo + educacao""", data = df)

In [218]:
sm.OLS(y, x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  renda   R-squared:                       0.300
Model:                            OLS   Adj. R-squared:                  0.295
Method:                 Least Squares   F-statistic:                     61.34
Date:                Sun, 12 Nov 2023   Prob (F-statistic):               0.00
Time:                        22:43:18   Log-Likelihood:            -1.5391e+05
No. Observations:               15000   AIC:                         3.080e+05
Df Residuals:                   14895   BIC:                         3.088e+05
Df Model:                         104                                         
Covariance Type:            nonrobust                                         
================================================================================================================
                                                   coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------------
Intercept                                      993.4636    648.167      1.533      0.125    -277.023    2263.950
C(idade, Treatment(40))[T.22]                 -252.0405   1824.321     -0.138      0.890   -3827.935    3323.854
C(idade, Treatment(40))[T.23]                -2118.0898   1400.333     -1.513      0.130   -4862.915     626.735
C(idade, Treatment(40))[T.24]                -1807.7050    768.826     -2.351      0.019   -3314.698    -300.712
C(idade, Treatment(40))[T.25]                -1474.5933    682.234     -2.161      0.031   -2811.856    -137.331
C(idade, Treatment(40))[T.26]                 -480.1998    608.194     -0.790      0.430   -1672.336     711.936
C(idade, Treatment(40))[T.27]                 -697.1013    454.125     -1.535      0.125   -1587.241     193.039
C(idade, Treatment(40))[T.28]                 -681.7077    460.768     -1.480      0.139   -1584.869     221.454
C(idade, Treatment(40))[T.29]                 -196.7231    472.408     -0.416      0.677   -1122.702     729.256
C(idade, Treatment(40))[T.30]                 -112.4092    453.947     -0.248      0.804   -1002.202     777.383
C(idade, Treatment(40))[T.31]                 -235.3653    463.671     -0.508      0.612   -1144.218     673.488
C(idade, Treatment(40))[T.32]                 -719.1419    446.529     -1.611      0.107   -1594.393     156.109
C(idade, Treatment(40))[T.33]                  386.7352    451.370      0.857      0.392    -498.005    1271.475
C(idade, Treatment(40))[T.34]                 -453.9526    461.772     -0.983      0.326   -1359.083     451.178
C(idade, Treatment(40))[T.35]                  573.6647    470.809      1.218      0.223    -349.178    1496.508
C(idade, Treatment(40))[T.36]                 -601.0745    469.152     -1.281      0.200   -1520.669     318.520
C(idade, Treatment(40))[T.37]                  256.5865    444.946      0.577      0.564    -615.563    1128.736
C(idade, Treatment(40))[T.38]                  -94.7378    450.443     -0.210      0.833    -977.661     788.185
C(idade, Treatment(40))[T.39]                  478.3155    448.812      1.066      0.287    -401.411    1358.042
C(idade, Treatment(40))[T.41]                 -772.7564    455.464     -1.697      0.090   -1665.522     120.010
C(idade, Treatment(40))[T.42]                  157.1535    461.568      0.340      0.734    -747.576    1061.883
C(idade, Treatment(40))[T.43]                 -560.6778    445.865     -1.258      0.209   -1434.629     313.273
C(idade, Treatment(40))[T.44]                 -249.1524    462.318     -0.539      0.590   -1155.353     657.048
C(idade, Treatment(40))[T.45]                 -389.7156    472.936     -0.824      0.410   -1316.729     537.297
C(idade, Treatment(40))[T.46]   

In [219]:
df = df[df["qt_pessoas_residencia"] >= 5]
max_qtd_filhos = df["qtd_filhos"].value_counts().idxmax()
max_idade = df["idade"].value_counts().idxmax()
max_tempo_emprego = df["tempo_emprego"].value_counts().idxmax()
max_qt_pessoas_residencia = df["qt_pessoas_residencia"].value_counts().idxmax()
max_qtd_filhos, max_idade, max_tempo_emprego, max_qt_pessoas_residencia

(3, 38, 4, 5)

In [220]:
y, x = patsy.dmatrices(f"""renda ~ C(idade, Treatment({max_idade}))
                        + C(qtd_filhos, Treatment({max_qtd_filhos}))
                        + C(qt_pessoas_residencia, Treatment({max_qt_pessoas_residencia}))
                        + educacao""", data = df)

sm.OLS(y, x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  renda   R-squared:                       0.258
Model:                            OLS   Adj. R-squared:                  0.149
Method:                 Least Squares   F-statistic:                     2.374
Date:                Sun, 12 Nov 2023   Prob (F-statistic):           0.000495
Time:                        22:43:18   Log-Likelihood:                -2047.1
No. Observations:                 205   AIC:                             4148.
Df Residuals:                     178   BIC:                             4238.
Df Model:                          26                                         
Covariance Type:            nonrobust                                         
================================================================================================================
                                                   coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------------
Intercept                                     2514.7241   1125.220      2.235      0.027     294.236    4735.212
C(idade, Treatment(38))[T.27]                 4150.7294   2086.668      1.989      0.048      32.939    8268.520
C(idade, Treatment(38))[T.28]                -1656.3586   4134.730     -0.401      0.689   -9815.755    6503.038
C(idade, Treatment(38))[T.29]                -1265.1356   3410.686     -0.371      0.711   -7995.718    5465.447
C(idade, Treatment(38))[T.30]                 2248.9170   2731.492      0.823      0.411   -3141.356    7639.190
C(idade, Treatment(38))[T.31]                 2805.9093   2273.215      1.234      0.219   -1680.010    7291.829
C(idade, Treatment(38))[T.32]                 1388.9936   4110.286      0.338      0.736   -6722.166    9500.153
C(idade, Treatment(38))[T.33]                 3613.3114   1696.033      2.130      0.035     266.392    6960.231
C(idade, Treatment(38))[T.34]                 1084.0132   1979.952      0.547      0.585   -2823.187    4991.213
C(idade, Treatment(38))[T.35]                 3606.5543   2167.593      1.664      0.098    -670.933    7884.042
C(idade, Treatment(38))[T.36]                 4381.3052   1589.040      2.757      0.006    1245.524    7517.086
C(idade, Treatment(38))[T.37]                 4934.3789   3103.489      1.590      0.114   -1189.987    1.11e+04
C(idade, Treatment(38))[T.39]                 2012.9696   1610.546      1.250      0.213   -1165.252    5191.191
C(idade, Treatment(38))[T.40]                -2151.2290   3410.686     -0.631      0.529   -8881.811    4579.353
C(idade, Treatment(38))[T.41]                 1117.9895   2046.278      0.546      0.586   -2920.096    5156.075
C(idade, Treatment(38))[T.42]                 1.491e+04   2217.346      6.726      0.000    1.05e+04    1.93e+04
C(idade, Treatment(38))[T.43]                 3262.2405   2739.815      1.191      0.235   -2144.459    8668.940
C(idade, Treatment(38))[T.44]                  823.2536   4110.286      0.200      0.841   -7287.906    8934.413
C(idade, Treatment(38))[T.45]                 1534.5060   2001.107      0.767      0.444   -2414.441    5483.453
C(idade, Treatment(38))[T.47]                 3638.4391   4449.326      0.818      0.415   -5141.775    1.24e+04
C(idade, Treatment(38))[T.48]                  110.9270   1380.940      0.080      0.936   -2614.193    2836.047
C(qtd_filhos, Treatment(3))[T.4]              9408.5933   3633.488      2.589      0.010    2238.338    1.66e+04
C(qtd_filhos, Treatment(3))[T.5]             -7196.7391   3368.241     -2.137      0.034   -1.38e+04    -549.918
C(qtd_filhos, Treatment(3))[T.7]             -2674.1915   1945.393     -1.375      0.171   -6513.194    1164.811
C(qtd_filhos, Treatment(3))[T.14

2. 
- O modelo melhorou, o AIC	foi de 3.100e+05 para 4148, ou seja o modelo com AIC de 4147 é preferível em termos de ajuste aos dados e complexidade em comparação com o modelo associado ao valor 3.100e+05.

- Os valores de qt_pessoas_residencia sofreram alterações com as adaptações do modelo, anteriormente seus P>|t| eram maiores do que no modelo ajustado.

antes: 
| Variável                                     | Coeficiente | Erro Padrão | Estatística T | Valor P | Intervalo de Confiança (95%) |
|---------------------------------------------|-------------|-------------|---------------|---------|------------------------------|
| C(qt_pessoas_residencia, Treatment(2))[T.1]  | -82.1195    | 159.116     | -0.516        | 0.606   | -394.007 a 229.768            |
| C(qt_pessoas_residencia, Treatment(2))[T.3]  | 173.6366    | 329.655     | 0.527         | 0.598   | -472.528 a 819.801            |
| C(qt_pessoas_residencia, Treatment(2))[T.4]  | 778.2655    | 910.267     | 0.855         | 0.393   | -1005.969 a 2562.500          |
| C(qt_pessoas_residencia, Treatment(2))[T.5]  | -1213.5376  | 3030.311    | -0.400        | 0.689   | -7153.321 a 4726.246          |
| C(qt_pessoas_residencia, Treatment(2))[T.6]  | 1589.2921   | 3910.218    | 0.406         | 0.684   | -6075.218 a 9253.802          |
| C(qt_pessoas_residencia, Treatment(2))[T.7]  | -1809.5989  | 6284.173    | -0.288        | 0.773   | -1.41e+04 a 1.05e+04         |
| C(qt_pessoas_residencia, Treatment(2))[T.9]  | -156.6511   | 1565.004    | -0.100        | 0.920   | -3224.253 a 2910.951         |
| C(qt_pessoas_residencia, Treatment(2))[T.15] | -2661.3827  | 2464.099    | -1.080        | 0.280   | -7491.320 a 2168.555         |


depois:
| Variável                                    | Coeficiente | Erro Padrão | Estatística T | Valor P | Intervalo de Confiança (95%) |
|--------------------------------------------|-------------|-------------|---------------|---------|------------------------------|
| C(qt_pessoas_residencia, Treatment(5))[T.6] | -6934.1215  | 3356.963    | -2.066        | 0.040   | -1.36e+04 a -309.555          |
| C(qt_pessoas_residencia, Treatment(5))[T.7] | 9145.9757   | 5446.225    | 1.679         | 0.095   | -1601.502 a 1.99e+04          |
| C(qt_pessoas_residencia, Treatment(5))[T.9] | -2674.1915  | 1945.393    | -1.375        | 0.171   | -6513.194 a 1164.811          |
| C(qt_pessoas_residencia, Treatment(5))[T.15]| 110.9270    | 1380.940    | 0.080         | 0.936   | -2614.193 a 2836.047          |



In [221]:
model = sm.OLS(y, x).fit()


In [222]:
idade_remover = [27, 30, 31, 33, 35, 36, 37, 39, 42, 43, 45, 47]
df = df[~df['idade'].isin(idade_remover)]

max_qtd_filhos = df["qtd_filhos"].value_counts().idxmax()
max_idade = df["idade"].value_counts().idxmax()
max_tempo_emprego = df["tempo_emprego"].value_counts().idxmax()
max_qt_pessoas_residencia = df["qt_pessoas_residencia"].value_counts().idxmax()
max_qtd_filhos, max_idade, max_tempo_emprego, max_qt_pessoas_residencia

y, x = patsy.dmatrices(f"""renda ~ C(idade, Treatment({max_idade}))
                        + C(qtd_filhos, Treatment({max_qtd_filhos}))
                        + C(qt_pessoas_residencia, Treatment({max_qt_pessoas_residencia}))
                        + educacao""", data = df)

sm.OLS(y, x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  renda   R-squared:                       0.161
Model:                            OLS   Adj. R-squared:                 -0.001
Method:                 Least Squares   F-statistic:                    0.9934
Date:                Sun, 12 Nov 2023   Prob (F-statistic):              0.463
Time:                        22:43:18   Log-Likelihood:                -626.50
No. Observations:                  69   AIC:                             1277.
Df Residuals:                      57   BIC:                             1304.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
================================================================================================================
                                                   coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------------
Intercept                                     3106.6314    537.067      5.784      0.000    2031.173    4182.089
C(idade, Treatment(38))[T.28]                -1268.0199   1727.124     -0.734      0.466   -4726.527    2190.487
C(idade, Treatment(38))[T.29]                -1530.2943   1417.210     -1.080      0.285   -4368.209    1307.620
C(idade, Treatment(38))[T.32]                 1287.2093   1703.351      0.756      0.453   -2123.692    4698.111
C(idade, Treatment(38))[T.34]                  732.7217    878.077      0.834      0.408   -1025.596    2491.040
C(idade, Treatment(38))[T.40]                -2416.3876   1417.210     -1.705      0.094   -5254.302     421.527
C(idade, Treatment(38))[T.41]                 1243.8904    850.109      1.463      0.149    -458.423    2946.204
C(idade, Treatment(38))[T.44]                  721.4693   1703.351      0.424      0.673   -2689.432    4132.371
C(idade, Treatment(38))[T.48]                  -86.3755    579.006     -0.149      0.882   -1245.816    1073.065
C(qtd_filhos, Treatment(3))[T.4]                98.8097    593.123      0.167      0.868   -1088.899    1286.518
C(qtd_filhos, Treatment(3))[T.14]              -86.3755    579.006     -0.149      0.882   -1245.816    1073.065
C(qt_pessoas_residencia, Treatment(5))[T.6]     98.8097    593.123      0.167      0.868   -1088.899    1286.518
C(qt_pessoas_residencia, Treatment(5))[T.15]   -86.3755    579.006     -0.149      0.882   -1245.816    1073.065
educacao[T.Superior completo]                 -270.3815    628.308     -0.430      0.669   -1528.546     987.783
educacao[T.Superior incompleto]                558.9874   1985.357      0.282      0.779   -3416.622    4534.597
==============================================================================
Omnibus:                       48.336   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              158.224
Skew:                           2.223   Prob(JB):                     4.39e-35
Kurtosis:                       8.939   Cond. No.                     1.04e+18
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 8.32e-35. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [223]:
idade_remover = [28, 29, 32, 34, 40, 41]
df = df[~df['idade'].isin(idade_remover)]

max_qtd_filhos = df["qtd_filhos"].value_counts().idxmax()
max_idade = df["idade"].value_counts().idxmax()
max_tempo_emprego = df["tempo_emprego"].value_counts().idxmax()
max_qt_pessoas_residencia = df["qt_pessoas_residencia"].value_counts().idxmax()
max_qtd_filhos, max_idade, max_tempo_emprego, max_qt_pessoas_residencia

y, x = patsy.dmatrices(f"""renda ~ C(idade, Treatment({max_idade}))
                        + C(qtd_filhos, Treatment({max_qtd_filhos}))
                        + C(qt_pessoas_residencia, Treatment({max_qt_pessoas_residencia}))
                        + educacao""", data = df)

sm.OLS(y, x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  renda   R-squared:                       0.018
Model:                            OLS   Adj. R-squared:                 -0.109
Method:                 Least Squares   F-statistic:                    0.1431
Date:                Sun, 12 Nov 2023   Prob (F-statistic):              0.965
Time:                        22:45:00   Log-Likelihood:                -328.10
No. Observations:                  36   AIC:                             666.2
Df Residuals:                      31   BIC:                             674.1
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
================================================================================================================
                                                   coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------------
Intercept                                     3249.5101    622.056      5.224      0.000    1980.819    4518.201
C(idade, Treatment(38))[T.44]                  683.5279   1727.735      0.396      0.695   -2840.210    4207.266
C(idade, Treatment(38))[T.48]                 -134.0017    595.315     -0.225      0.823   -1348.155    1080.151
C(qtd_filhos, Treatment(3))[T.4]              -403.3151   1223.953     -0.330      0.744   -2899.584    2092.954
C(qtd_filhos, Treatment(3))[T.14]             -134.0017    595.315     -0.225      0.823   -1348.155    1080.151
C(qt_pessoas_residencia, Treatment(5))[T.6]   -403.3151   1223.953     -0.330      0.744   -2899.584    2092.954
C(qt_pessoas_residencia, Treatment(5))[T.15]  -134.0017    595.315     -0.225      0.823   -1348.155    1080.151
educacao[T.Superior completo]                 -480.2561    827.925     -0.580      0.566   -2168.820    1208.308
==============================================================================
Omnibus:                       43.026   Durbin-Watson:                   2.327
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              150.641
Skew:                           2.815   Prob(JB):                     1.94e-33
Kurtosis:                      11.290   Cond. No.                     9.82e+18
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 4.92e-37. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

3. O modelo parece estar melhor considerando o AIC que está em 666.2 mesmo o R-squared sendo 0.018